certcache.go

```diff
@@ -16,24 +16,74 @@ package amppackager
 
 import (
 	"bytes"
+	"context"
 	"crypto/x509"
+	"encoding/base64"
+	"io"
+	"io/ioutil"
+	"log"
 	"net/http"
+	"net/url"
+	"sync"
 	"time"
 
 	"github.com/WICG/webpackage/go/signedexchange/certurl"
 	"github.com/julienschmidt/httprouter"
+	"github.com/pkg/errors"
+	"github.com/pquerna/cachecontrol"
+	"golang.org/x/crypto/ocsp"
 )
 
+// The maximum representable time, per https://stackoverflow.com/questions/25065055/what-is-the-maximum-time-time-in-go.
+var infiniteFuture = time.Unix(1<<63-62135596801, 999999999)
+
+// The OCSP code below aims to satisfy items 1-8 of the "sleevi" doc:
+// https://gist.github.com/sleevi/5efe9ef98961ecfb4da8
+// Item #9 should be added. Item #10 is N/A for SXGs.
+
+// 1MB is the maximum used by github.com/xenolf/lego/acmev2 in GetOCSPForCert.
+// Alternatively, here's a random documented example of a 20K limit:
+// https://www.ibm.com/support/knowledgecenter/en/SSPREK_9.0.0/com.ibm.isam.doc/wrp_stza_ref/reference/ref_ocsp_max_size.html
+const maxOCSPResponseBytes = 1024 * 1024
+
+// How often to check if OCSP stapling needs updating.
+const ocspCheckInterval = 1 * time.Hour
+
+var fakeOCSPServer string
```

Comment:
"fake" in code? code smell?

Reply:
Totally a code smell. I added a comment and am updating these to be struct fields (work-in-progress) rather than globals, but I don't see a way around them.

OCSP server is normally extracted from the certificate, which is static [1], but we need to point it at a fake server, which gets a dynamically-assigned port from the httptest library.

OCSP expiry is normally computed from the HTTP response headers, by the 3rd party cachecontrol library. That library uses `time.Now` internally, so we need to stub out our use of it for test.

[1] I guess we could mint certs in test setup, but eww.